Formal2024/Workshops/workshop2.lean

```diff
@@ -0,0 +1,239 @@
+import Mathlib.Data.Real.Basic
+import Mathlib.Tactic
+
+namespace Week2
+
+/-
+  # Logic: Implication, universal quantifier, existential quantifier, negation
+  References: [MIL] Mathematics in Lean, [Tut] Tutorials project.
+  Most of the demonstration section comes [MIL].
+-/
+
+/-!
+  # Implication and universal quantifier
+-/
+
+-- To write `∀` type
+
+-- Lean treats the universal quantifier as an implication.
+-- Implications are grouped on the RIGHT (N.B. this is different from the usual convention)
+
+theorem my_lemma : ∀ x y ε : ℝ, 0 < ε → ε ≤ 1 → |x| < ε → |y| < ε → |x * y| < ε :=
+  sorry
+
+-- Lean is all about types, and checking that something is of the right type
+section
+-- Naively, you can think of types as sets. Here these variables are real numbers
+variable (a b δ : ℝ)
+-- Being of the type of a Proposition means being a proof
+variable (h₀ : 0 < δ) (h₁ : δ ≤ 1)
+variable (ha : |a| < δ) (hb : |b| < δ)
+
+-- An implication is something like a function from sets of proofs to set of proofs
+#check my_lemma a b δ
+#check my_lemma a b δ h₀ h₁
+#check my_lemma a b δ h₀ h₁ ha hb
+
+end
+
+
+-- To prove an implication statement, we use intro
+
+example: ∀ x y z : ℝ, x ≤ y → y ≤ z → x ≤ z := by
+  intro x y z h₀ h₁
+  -- Now we can proceed as usual
+  apply le_trans
+  · apply h₀
+  · apply h₁
+
+
+-- Using the universal quantifier, we define an upper bounded function
+def FnUb (f : ℝ → ℝ) (a : ℝ) : Prop :=
+  ∀ x, f x ≤ a
+
+
+
+variable (f g : ℝ → ℝ) (a b : ℝ)
+
+-- It is useful to define functions, this can be done with the command `fun`
+
+#check fun x ↦ x + 1
+#check fun x ↦ f x + g x
```

Comment:
We should add examples of the other equivalent notations: `fun x => x + 1`, `λ x => x + 1` and `λ x ↦ x + 1`.

Reply:
This is one of the things I was unsure of. I feel like there's already a lot to cover in this session, so I sticked to a single notation when there were many available (for example I skipped all the discussion about obtain, cases, etc... that is in MIL). So let me know what you all think about this

Reply:
I thought the lambda function was completely replaced by `fun` in Lean 4?

Reply:
Apparently you can still use `\lambda` as a shorthand

Reply:
oh, but you can't do `λ x, x + 1`, right? that's only for Lean 3? (as far as I know)